File: rust/src/detectors/type_mismatch.rs
use crate::detectors::{Definition, Detector, IndexedModel, def_sort_key, make_finding};
use crate::models::{Finding, Severity, VariableType};
use std::collections::HashMap;

/// Type mismatch: the same variable is defined with values of incompatible
/// inferred types across environment files (e.g. `PORT=3000` in development
/// but `PORT=3000abc` in production). The "expected" type is taken from the
/// development file when present, otherwise the most common type. Only
/// variable *types* and locations are reported — never values.
pub struct TypeMismatchDetector;

impl Detector for TypeMismatchDetector {
    fn id(&self) -> &'static str {
        "type-mismatch"
    }

    fn name(&self) -> &'static str {
        "type-mismatch"
    }

    fn description(&self) -> &'static str {
        "The same variable has incompatible inferred types across environment files."
    }

    fn detect(&self, index: &IndexedModel) -> Vec<Finding> {
        let mut findings = Vec::new();

        let mut entries: Vec<(&String, &Vec<Definition>)> = index.env_definitions.iter().collect();
        entries.sort_by(|(na, da), (nb, db)| def_sort_key(da).cmp(&def_sort_key(db)).then(na.cmp(nb)));

        for (name, defs) in entries {
            let typed: Vec<&crate::detectors::Definition> = defs
                .iter()
                .filter(|d| d.value.is_some() && d.var_type != VariableType::Unknown && !d.value.as_ref().unwrap().is_empty())
                .collect();
            if typed.len() < 2 {
                continue;
            }

            let distinct_types: std::collections::HashSet<_> = typed.iter().map(|d| d.var_type).collect();
            if distinct_types.len() < 2 {
                continue;
            }

            let expected = typed
                .iter()
                .find(|d| d.environment.as_deref() == Some("development"))
                .map(|d| d.var_type)
                .unwrap_or_else(|| most_common_type(&typed));

            for def in typed {
                if def.var_type == expected {
                    continue;
                }
                findings.push(make_finding(
                    "type-mismatch",
                    Severity::Error,
                    name,
                    format!("expected: {}, found: {}", expected, def.var_type),
                    vec![def.origin.clone()],
                ));
            }
        }

        findings
    }
}

fn most_common_type(defs: &[&crate::detectors::Definition]) -> VariableType {
    // Count occurrences but resolve ties by first-seen order, matching the
    // reference CLI which iterates its insertion-ordered `Map`.
    let mut counts: HashMap<VariableType, usize> = HashMap::new();
    let mut order: Vec<VariableType> = Vec::new();
    for d in defs {
        if !counts.contains_key(&d.var_type) {
            order.push(d.var_type);
        }
        *counts.entry(d.var_type).or_default() += 1;
    }
    let mut best = order[0];
    let mut best_count = 0;
    for ty in order {
        let count = counts[&ty];
        if count > best_count {
            best = ty;
            best_count = count;
        }
    }
    best
}

File: dart/lib/src/detectors/type_mismatch.dart
import '../models/finding.dart';
import '../models/variable_type.dart';
import 'detector.dart';

/// Type mismatch: the same variable is defined with values of incompatible
/// inferred types across environment files. The "expected" type is taken from
/// the development file when present, otherwise the most common type. Only
/// variable *types* and locations are reported — never values.
class TypeMismatchDetector implements Detector {
  @override
  String get id => 'type-mismatch';

  @override
  String get name => 'type-mismatch';

  @override
  String get description =>
      'The same variable has incompatible inferred types across environment files.';

  @override
  List<Finding> detect(IndexedModel index) {
    final findings = <Finding>[];

    for (final entry in sortedEntries(index.envDefinitions, defSortKey)) {
      final typed = entry.value
          .where((d) =>
              d.value != null &&
              d.type != VariableType.unknown &&
              d.value!.isNotEmpty)
          .toList();
      if (typed.length < 2) continue;

      final distinctTypes = typed.map((d) => d.type).toSet();
      if (distinctTypes.length < 2) continue;

      VariableType? devType;
      for (final d in typed) {
        if (d.environment == 'development') {
          devType = d.type;
          break;
        }
      }
      final resolvedExpected = devType ?? _mostCommonType(typed);

      for (final def in typed) {
        if (def.type == resolvedExpected) continue;
        findings.add(Finding.make(
          'type-mismatch',
          Severity.error,
          entry.key,
          'expected: ${resolvedExpected.str}, found: ${def.type.str}',
          [def.origin],
        ));
      }
    }

    return findings;
  }

  static VariableType _mostCommonType(List<Definition> defs) {
    // Count occurrences but resolve ties by first-seen order, matching the
    // reference CLI.
    final counts = <VariableType, int>{};
    final order = <VariableType>[];
    for (final d in defs) {
      if (!counts.containsKey(d.type)) order.add(d.type);
      counts[d.type] = (counts[d.type] ?? 0) + 1;
    }
    var best = order.first;
    var bestCount = -1;
    for (final ty in order) {
      final count = counts[ty]!;
      if (count > bestCount) {
        best = ty;
        bestCount = count;
      }
    }
    return best;
  }
}
